### ISSUE TITLE
Update communication backend to send/recv header + timestamped data

### ISSUE THREAD

**Issue by BenjaminRodenberg** [precice/precice#1957]
In https://github.com/precice/precice/pull/1664 I introduced bulk send/recv of multiple substeps. The main motivation is to support multirate time stepping and higher-order interpolation. This, however, came with some technical debt, because I did not consistently update the communication backend. Instead, I decided to use the existing backend with three stages of communication:

1. send/recv number of exchanged timesteps `n` (1 int)
2. send/recv array of timestamps (`n` doubles)
3. send/recv array of actual data (`n * dim * size` doubles)

Note that we have to know `n`, before we can trigger the second and the third step.

A more appropriate approach would be the following:

1. send/recv a header with the number of exchanged timesteps `n`
2. send/recv the actual data + timestamps

This approach should also help us to remove the pretty confusing function `BaseCouplingScheme::receiveDataForWindowEnd`, because we would by default always attach a timestamp to the data and, therefore, we would not need to call `getTime()` inside `receiveData` to implicitly derive the time of the received data.

**Comment by fsimonis** [comment]
While this is a good idea overall, this is out of scope for the v3.2.0 release. Thanks to #2208 it is also not a massive performance hurdle anymore.
